### PR TITLE
HAProxy Router: Don't health-check idled services

### PR DESCRIPTION
**Pull request by DirectXMan12**
Previously, all "endpoints" were health-checked by HAProxy.  For
backends which connected via non-secured connections, this entailed
partial establishment of a TCP connection by HAProxy, which did not
trigger the unidler proxy.  However, for secured backends, the health
check involved attempting to start a TLS connection, which used a full
TCP connection, which caused a wakeup to occur.

This commit adds an extra field to the template data's endpoint
structure which indicates that a particular endpoint is to be considered
an "idled" endpoint, meaning that no health check is necessary.  The
default template then uses this to disable health checking on the given
backend endpoints, which actually point to service IPs backed by the
unidling proxy.

Fixes bug 1366180

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @knobunc

**Comment by knobunc** [comment]
LGTM

**Comment by sjenning** [comment]
lgtm

File: pkg/router/template/plugin.go [hunk 1]
```diff
@@ -311,6 +313,8 @@ func createRouterEndpoints(endpoints *kapi.Endpoints, excludeUDP bool, lookupSvc
 					Port: strconv.Itoa(int(p.Port)),
 
 					PortName: p.Name,
+
+					Idled: wasIdled,
```
Comment (smarterclayton) [inline pkg/router/template/plugin.go]: Could we just make this `NoHealthCheck` so the template doesn't have to know about it?

Comment (DirectXMan12) [inline pkg/router/template/plugin.go]: sure

**Comment by ncdc** [comment]
Is it possible to have a test case for this?

**Comment by DirectXMan12** [comment]
I may be able to put in an extended test case.  Let me see.

**Comment by DirectXMan12** [comment]
Tweaked the extended test to test this.

**Comment by knobunc** [comment]
[merge]

**Comment by ncdc** [comment]
@DirectXMan12 rebase needed

**Comment by knobunc** [comment]
[merge]

File: images/router/haproxy/conf/haproxy-config.template [hunk 2]
```diff
@@ -291,17 +291,21 @@ backend be_edge_http_{{$cfgIdx}}
       {{ if ne $weight 0 }}
         {{ with $serviceUnit := index $.ServiceUnits $serviceUnitName }}
           {{ range $idx, $endpoint := endpointsForAlias $cfg $serviceUnit }}
-            {{ with $healthIntv := index $cfg.Annotations "router.openshift.io/haproxy.health.check.interval" }}
-              {{ if (matchPattern "[1-9][0-9]*(us|ms|s|m|h|d)?" $healthIntv) }}
+            {{ if no $endpoint.NoHealthCheck }}
```
Comment (ncdc) [inline images/router/haproxy/conf/haproxy-config.template]: F0818 18:30:04.056833       1 helpers.go:108] error: template: haproxy-config.template:294: function "no" not defined

File: images/router/haproxy/conf/haproxy-config.template [hunk 3]
```diff
@@ -334,17 +338,21 @@ backend be_tcp_{{$cfgIdx}}
...
```
Comment (ncdc) [inline images/router/haproxy/conf/haproxy-config.template]: F0818 18:30:04.056833       1 helpers.go:108] error: template: haproxy-config.template:294: function "no" not defined

File: images/router/haproxy/conf/haproxy-config.template [hunk 4]
```diff
@@ -382,19 +390,23 @@ backend be_secure_{{$cfgIdx}}
       {{ if ne $weight 0 }}
         {{ with $serviceUnit := index $.ServiceUnits $serviceUnitName }}
           {{ range $idx, $endpoint := endpointsForAlias $cfg $serviceUnit }}
-            {{ with $healthIntv := index $cfg.Annotations "router.openshift.io/haproxy.health.check.interval" }}
-              {{ if (matchPattern "[1-9][0-9]*(us|ms|s|m|h|d)?" $healthIntv) }}
-  server {{$endpoint.IdHash}} {{$endpoint.IP}}:{{$endpoint.Port}} ssl check inter {{$healthIntv}} verify required ca-file {{ $workingDir }}/cacerts/{{$cfgIdx}}.pem cookie {{$endpoint.IdHash}} weight {{$weight}}
-              {{ else }}
-  server {{$endpoint.IdHash}} {{$endpoint.IP}}:{{$endpoint.Port}} ssl check inter 5000ms verify required ca-file {{$workingDir}}/cacerts/{{$cfgIdx}}.pem cookie {{$endpoint.IdHash}} weight {{$weight}}
-              {{ end }}
+            {{ if no $endpoint.NoHealthCheck }}
```
Comment (ncdc) [inline images/router/haproxy/conf/haproxy-config.template]: F0818 18:30:04.056833       1 helpers.go:108] error: template: haproxy-config.template:294: function "no" not defined

**Comment by ncdc** [comment]
@knobunc green jenkins

**Comment by knobunc** [comment]
LGTM [merge]

**Pull request merged**